app/helpers/locale_helper.rb

```diff
@@ -10,7 +10,17 @@ def with_user_locale(user, &block)
     locale = user.email_language
 
     if I18n.locale_available?(locale)
-      I18n.with_locale(locale, &block)
+      # Handle url_helper's url_options
+      if defined?(url_options)
+        current_locale = I18n.locale
+        I18n.with_locale(locale) do
+          url_options[:locale] = locale
+          block.call
+          url_options[:locale] = current_locale
```

Comment:
A/N: So this won't work as `url_options` is frozen.

```ruby
FrozenError: can't modify frozen Hash: {:host=>"test.host", :port=>nil, :protocol=>"http://", :_recall=>{}}

  0) LocaleHelper#with_user_locale when the user has an email_language sets the language inside the block and yields
     Failure/Error: url_options[:locale] = locale

     FrozenError:
       can't modify frozen Hash: {:host=>"test.host", :port=>nil, :protocol=>"http://", :_recall=>{}}
```

Reply:
We could override the `url_options` method to create a new copy of the hash with merged properties.

```diff
diff --git a/app/helpers/locale_helper.rb b/app/helpers/locale_helper.rb
index 90559fa92d..d4e4007eff 100644
--- a/app/helpers/locale_helper.rb
+++ b/app/helpers/locale_helper.rb
@@ -2,4 +2,8 @@
 
 module LocaleHelper
+  def url_options
+    super.merge(locale: I18n.locale)
+  end
+
   def locale_url_param
     active_locale = I18n.locale
@@ -11,15 +15,5 @@ module LocaleHelper
 
     if I18n.locale_available?(locale)
-      # Handle url_helper's url_options
-      if defined?(url_options)
-        current_locale = I18n.locale
-        I18n.with_locale(locale) do
-          url_options[:locale] = locale
-          block.call
-          url_options[:locale] = current_locale
-        end
-      else
-        I18n.with_locale(locale, &block)
-      end
+      I18n.with_locale(locale, &block)
     else
       Rails.logger.warn("user_id=#{user.uuid} has bad email_language=#{locale}") if locale.present?
```

Reply:
@jmhooper and I think this is Rspec shenanigans. After updating the LocaleHelper spec,  I'm no longer seeing the error.

Are you comfortable with the implementation as it stands?

Reply:
Fine with me either way 👍